### DISCUSSION TITLE
client "delete" method missing "data" and "json" parameters

### DISCUSSION THREAD

**Discussion by matletix** [encode/httpx#1587] [Potential Issue] [answered]
Hi,

I'm using the AsyncClient for making API calls to an external API.
I need to make DELETE calls with a JSON body (to specify a list of values to be deleted), but the 'json' or 'data' parameters are not present in the `delete` method definition (while they are present in the POST method for example).
For now I'm using the `client.request` method with the `method='DELETE'` parameter, and the 'json' parameter is available in this method.
Is this on purpose ? Can we consider adding the data and json parameters to the detele method of the client ?

**Comment by lovelydinosaur** [comment, marked as answer]
> Is this on purpose?

Yes.

> Can we consider adding the data and json parameters to the detele method of the client?

Probably not, no.

See our docs at https://www.python-httpx.org/compatibility/#request-body-on-http-methods and the HTTP RFC https://tools.ietf.org/html/rfc7231#section-4.3.5

> A payload within a DELETE request message has no defined semantics;
> sending a payload body on a DELETE request might cause some existing
> implementations to reject the request.

So, `DELETE` requests with a body are not completely disallowed, but it's discouraged as a usage. For this reason we don't support it in the general `.delete()` API, but we do still allow for it if that's *really* what you intended.

**Reply by matletix** [reply]
Ok, thank you for your answer. So using the `request` method instead of the `delete` method is my best option to accomplish this, right ?

**Reply by lovelydinosaur** [reply]
Yup.

**Reply by shurshilov** [reply]
This is a mistake. What if I need to delete a list of IDs that has a long length, I can't put these parameters in the url because it is limited. All libraries allow you to transfer json body inside delete method.

**Reply by deanm0000** [reply]
@shurshilov As tom said `httpx` allows it but only through the `request` method not the `delete` method since it's discouraged. That's how you'd delete a list of IDs with a long length.
